Public calendar

matthiasmoritz opened the pull request:
This Feature allows users to provid the own Calendar to all (and even anonymus) visitors.

It's my first pice of code for friendica, and my first pull request to a open source project so code quality might be not perfect ;)

matthiasmoritz commented:
By the way, is there any tutorial how to execute the checks?

nupplaphil commented on src/Model/Event.php:
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
+
+	public static function getPublicById( int $event_id ): array
+	{
+		/* To prevent from requesting private events via api we must add something like
+		 * $sql_perms = $sql_extra = and allow_cid is null and allow_gid is null and deny_cid is null and deny_gid is null
+		 * but I guess there is a methode alredy implemented for this?
+		 */
```
Please move this part to a PHP Doc header

nupplaphil commented on src/Model/Event.php:
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
+		 * $sql_perms = $sql_extra = and allow_cid is null and allow_gid is null and deny_cid is null and deny_gid is null
+		 * but I guess there is a methode alredy implemented for this?
+		 */
+		$events = self::getById( $event_id, '' );
```
```suggestion
		$events = self::getById($event_id, '');
```

nupplaphil commented on src/Model/Event.php:
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
+		return $events[0];
+	}
+
+	private static function getById ( int $event_id, string $additionalCondition) : array
```
```suggestion
	private static function getById (int $event_id, string $additionalCondition) : array
```

nupplaphil commented on src/Module/Calendar/Event/Get.php:
```diff
@@ -52,18 +54,23 @@ public function __construct(L10n $l10n, App\BaseURL $baseUrl, App\Arguments $arg
-		}
-
-		System::jsonExit($events ? self::map($events) : []);
+			if (!$this->session->getLocalUserId() && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
+				throw new HTTPException\UnauthorizedException();
+			}
+	
+			// get events by id or by date
+			if (!empty($request['id']) && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
+				$events = [Event::getByIdAndUid($this->session->getLocalUserId(), $request['id'], $this->parameters['nickname'] ?? null)];
+			} elseif (!empty($request['id']) && Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
+				$events = [Event::getPublicById($request['id'])];
+			} else {
+				$events = Event::getListByDate($this->session->getLocalUserId(), $request['start'] ?? '', $request['end'] ?? '', false, $this->parameters['nickname'] ?? null);
+			}
+	
+			System::jsonExit($events ? self::map($events) : []);
```
```suggestion
		if (!$this->session->getLocalUserId() && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			throw new HTTPException\UnauthorizedException();
		}

		// get events by id or by date
		if (!empty($request['id']) && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			$events = [Event::getByIdAndUid($this->session->getLocalUserId(), $request['id'], $this->parameters['nickname'] ?? null)];
		} elseif (!empty($request['id']) && Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			$events = [Event::getPublicById($request['id'])];
		} else {
			$events = Event::getListByDate($this->session->getLocalUserId(), $request['start'] ?? '', $request['end'] ?? '', false, $this->parameters['nickname'] ?? null);
		}

		System::jsonExit($events ? self::map($events) : []);
```

nupplaphil commented on src/Module/Calendar/Event/Get.php:
```diff
@@ -52,18 +54,23 @@ public function __construct(L10n $l10n, App\BaseURL $baseUrl, App\Arguments $arg
-			throw new HTTPException\UnauthorizedException();
-		}
+		$calenderOwner= User::getOwnerDataByNick(!empty($this->parameters['nickname']) ? $this->parameters['nickname'] : '');
+		//if (Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar') && !empty($request['id']) ){
```
Can you delete this if it isn't necessary anymore?

MrPetovan commented:
The checks are meant to be run by GitHub when you submit a PR and any time you push the branch. If you want to perform these checks in advance of the review, you can open a Draft PR (or convert this one to Draft) so that there's no pressure to add more changes to it.

nupplaphil commented:
Thank you for your help!! :)

Additionally, can you:
- Rebase this branch onto the latest `develop`. This will fix some checks
- Recreate the `view/lang/C/messages.po` --> plz execute the `bin/run_xgettext.sh` script to do that

MrPetovan commented:
Thank you for your work! For consistency, Friendica currently enforces the PSR-2 coding standards, see https://www.php-fig.org/psr/psr-2/

I've noticed too many violations in your changes to individually point them out, would like to take a first pass or for me to take over the branch?

matthiasmoritz commented:
@MrPetovan let me first go through the code. There is also still one functional issue that can cause privacy errors. I'll update the PR.

matthiasmoritz commented:
@MrPetovan can you please take over the PR. The Feature is working as excepted but I'm not shure how to continiue with the rest.

MrPetovan commented:
I will, thank you for your work!

MrPetovan commented:
Oh Lord, I'm diving down the calendar rabbit hole and nothing about the whole feature seems to make sense (not your specific changes) and I'm rewriting way more than I should but this is the life that chose me.

matthiasmoritz commented:
Sorry for opening pandoras box 😬

Quix0r reviewed (approved):
Type-hint checks are fine so far.

MrPetovan committed:
Remove unused theme info value "events_in_profile"
All public events (event if they are ownded by other users) are visible

MrPetovan committed:
Tighten profile restriction feature

- Prevent feed access to restricted profiles
- Rework display of restricted profiles with a redirect to the profile/restricted route
- Normalize permission checking with IHandleUserSession->isAuthenticated
- Remove unusable "nocache" parameter in feed module because session isn't initialized
- Reword setting name and description

MrPetovan committed:
Remove unused first parameter from BaseProfile::getTabsHTML

- Add expected Profile::load call to Profile\Photos module

MrPetovan committed:
Add new public_calendar additional feature

- This gives anonymous access to public events

MrPetovan committed:
Update main translation file after changing strings

MrPetovan commented:
Alright! This took me on two separate tangents (the theme variables and the `hidewall` setting) before getting to the main issue which is event retrieval and permission check, but it's there.

@matthiasmoritz You may be surprised to find very little of your own changes in the final PR, this is because I've simplified the existing `Model\Event` methods, which made many of your changes unnecessary.

Overall, there was a lot of unused/confusing code I removed and it's always a good thing.

The `hidewall` setting has been normalized (it was inconsistent in checking for authentication) and now covers a few more endpoints, including syndication feeds, so I've edited the setting description to reflect this new state of affairs, hopefully it's clearer.

Quix0r reviewed (approved):
Type-hint check okay.

nupplaphil commented:
Here we go :-) LGTM

MrPetovan commented:
@matthiasmoritz Thank you for your work again! You've helped improve Friendica on three separate fronts!

matthiasmoritz commented:
@MrPetovan  Not so sure how my three lines of code contributed to your work :) but anyway, I can test the feature and i think you broke something in the UI. See #12308
![image](https://user-images.githubusercontent.com/1984982/205140286-f51593ef-f69b-4868-806f-db910de7fd7c.png)

MrPetovan commented:
Are you sure it was working before? Do you want to look into it? This display comes from the `Module\Event\Show` class.

I've noticed we were performing a `Strings::escapeHTML` before a `BBCode::convert` so this may be the result of a double-encoding issue.

matthiasmoritz commented:
It works on stable 🤷‍♂️

MrPetovan commented:
So it could be a regression. I'll look into it if you don't want to.
